keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -433,6 +463,107 @@ class SqlArtifactRepository(
     }
   }
 
+  override fun pinEnvironment(deliveryConfig: DeliveryConfig, environmentArtifactPin: EnvironmentArtifactPin) {
+    with(environmentArtifactPin) {
+      val environment = deliveryConfig.environmentNamed(targetEnvironment)
+      val artifact = get(deliveryConfigName, reference, ArtifactType.valueOf(type.toUpperCase()))
+
+      sqlRetry.withRetry(WRITE) {
+        jooq.insertInto(ENVIRONMENT_ARTIFACT_PIN)
+          .set(ENVIRONMENT_ARTIFACT_PIN.ENVIRONMENT_UID, deliveryConfig.getUidFor(environment))
+          .set(ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_UID, artifact.uid)
+          .set(ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_VERSION, version)
+          .set(ENVIRONMENT_ARTIFACT_PIN.PINNED_AT, clock.millis())
+          .set(ENVIRONMENT_ARTIFACT_PIN.PINNED_BY, pinnedBy ?: "anonymous")
+          .set(ENVIRONMENT_ARTIFACT_PIN.COMMENT, comment)
+          .execute()
+      }
+    }
+  }
+
+  override fun pinnedEnvironments(deliveryConfig: DeliveryConfig): List<PinnedEnvironment> {
+    return sqlRetry.withRetry(READ) {
+      jooq.select(
+        ENVIRONMENT.NAME,
+        ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_VERSION,
+        DELIVERY_ARTIFACT.NAME,
+        DELIVERY_ARTIFACT.TYPE,
+        DELIVERY_ARTIFACT.DETAILS,
+        DELIVERY_ARTIFACT.REFERENCE
+      )
+        .from(ENVIRONMENT)
+        .innerJoin(ENVIRONMENT_ARTIFACT_PIN)
+        .on(ENVIRONMENT_ARTIFACT_PIN.ENVIRONMENT_UID.eq(ENVIRONMENT.UID))
+        .innerJoin(DELIVERY_ARTIFACT)
+        .on(DELIVERY_ARTIFACT.UID.eq(ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_UID))
+        .innerJoin(DELIVERY_CONFIG)
+        .on(DELIVERY_CONFIG.UID.eq(ENVIRONMENT.DELIVERY_CONFIG_UID))
+        .where(DELIVERY_CONFIG.NAME.eq(deliveryConfig.name))
+        .fetch { (environmentName, version, artifactName, type, details, reference) ->
+          PinnedEnvironment(
+            deliveryConfigName = deliveryConfig.name,
+            targetEnvironment = environmentName,
+            artifact = mapToArtifact(
+              artifactName,
+              ArtifactType.valueOf(type.toUpperCase()),
+              details,
+              reference,
+              deliveryConfig.name),
+            version = version)
+        }
+    }
+  }
+
+  override fun deletePin(deliveryConfig: DeliveryConfig, targetEnvironment: String) {
+    sqlRetry.withRetry(WRITE) {
+      jooq.select(ENVIRONMENT_ARTIFACT_PIN.ENVIRONMENT_UID, ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_UID)
+        .from(ENVIRONMENT_ARTIFACT_PIN)
+        .innerJoin(ENVIRONMENT)
+        .on(ENVIRONMENT_ARTIFACT_PIN.ENVIRONMENT_UID.eq(ENVIRONMENT.UID))
+        .where(
+          ENVIRONMENT.NAME.eq(targetEnvironment),
+          ENVIRONMENT.DELIVERY_CONFIG_UID.eq(deliveryConfig.uid))
+        .fetch { (envUid, artUid) ->
+          deletePin(envUid, artUid)
```

Comment:
noob sql question: do we need transactions here, or is this already going to happen together? or no need to worry about that since it's a read then a write?

Reply:
There's no advantage to including the select in a txn (no impact on behavior, we run with a `READ_COMMITTED`, not `REPEATABLE_READ` txn isolation level) and I want to avoid a transaction around the deletes in the scenario where an environment has multiple artifacts pinned.